[#1976] Move hot state management into store

mbaxter opened the pull request:
<!-- Thanks for sending a pull request! Please check out our contribution guidelines: -->
<!-- https://github.com/PegaSysEng/teku/blob/master/CONTRIBUTING.md -->

## PR Description
As a preliminary step in addressing #1976, move hot state management into the `Store`.  This consolidates hot state management in one place:  previously hot states were kept in both the `Database` and the `Store`.  This change also necessitated moving pruning logic from the `Database` into the `Store`.

Additionally, all knowably non-canonical blocks are now pruned whenever we finalized a new checkpoint.  This includes blocks that do not descend from the latest finalized block and blocks that descend from the latest finalized block, but are at or prior to latest finalized epoch boundary.   For example, if block 30 finalizes epoch 1 (with epoch boundary at slot 32), other blocks at slots 30, 31, or 32 will be pruned as invalid. 

As a result of these changes, a fair number of tests needed to be updated to operate on valid chain data.

mbaxter committed:
Rework Database state api - only return finalized states

mbaxter committed:
Move hotState managment and pruning logic into RecentChainData

Make pruning more robust - prune all knowably non-canonical blocks

mbaxter committed:
Clean up

ajsutton commented on ethereum/core/src/main/java/tech/pegasys/teku/core/StateTransition.java:
```diff
@@ -62,6 +70,60 @@ public StateTransition(BlockValidator blockValidator) {
     this(Optional.empty(), blockValidator);
   }
 
+  public static Map<Bytes32, BeaconState> produceStatesForBlocks(
```
This doesn't feel like the right place for this.  I'd be very tempted to split it into it's own class.  My natural inclination is to make it not static as well but given where it gets used from maybe not in this case...

mbaxter replied:
Moved this into a new class called `StateGenerator`

ajsutton commented on storage/src/main/java/tech/pegasys/teku/storage/Store.java:
```diff
@@ -354,15 +450,68 @@ public VoteTracker getVote(UnsignedLong validatorIndex) {
+
+      final Set<Bytes32> prunedRootsSet =
+          prunedHotBlockRoots.values().stream()
+              .flatMap(roots -> roots.stream())
```
nit: Can use a method reference here.

ajsutton commented on storage/src/main/java/tech/pegasys/teku/storage/Store.java:
```diff
@@ -354,15 +450,68 @@ public VoteTracker getVote(UnsignedLong validatorIndex) {
     public SafeFuture<Void> commit() {
+      final Map<Bytes32, SignedBeaconBlock> hotBlocks = new HashMap<>();
+      // To start all blocks are assumed to be hot blocks
+      hotBlocks.putAll(blocks);
```
nit: Can just be:
```suggestion
      // To start all blocks are assumed to be hot blocks
      final Map<Bytes32, SignedBeaconBlock> hotBlocks = new HashMap<>(blocks);
```

ajsutton commented on storage/src/main/java/tech/pegasys/teku/storage/Store.java:
```diff
@@ -354,15 +450,68 @@ public VoteTracker getVote(UnsignedLong validatorIndex) {
 
     @CheckReturnValue
     public SafeFuture<Void> commit() {
```
This method has gotten pretty big.  It's not easy to split up but I wonder if the calculation of finalised chain data could be separated out and return an object with `newlyFinalizedBlock, finalizedChainData, prunedHotBlockRoots, staleCheckpointStates`.  It's probably more code overall but I think helps to improve readability overall though.

Happy to leave it for now if it's likely to get changed again in later PRs though.

mbaxter replied:
Yeah - there's a lot of logic here.  Tried to reorganize a little here, but not sure how much better this is: https://github.com/PegaSysEng/teku/pull/2006/commits/6199f6dff80b7815eb2b6f1354e4e010ca89bb03

It's messy because there are all of the cross-references to the current transaction and underlying `Store`.

ajsutton commented on storage/src/main/java/tech/pegasys/teku/storage/Store.java:
```diff
@@ -403,6 +561,74 @@ public VoteTracker getVote(UnsignedLong validatorIndex) {
+      // Build combined index from slot to block root for new and existing blocks
+      final NavigableMap<UnsignedLong, Set<Bytes32>> slotToHotBlockRootIndex = new TreeMap<>();
+      indexBlockRootsBySlot(slotToHotBlockRootIndex, newBlocks.values());
+      slotToHotBlockRootIndex.putAll(rootsBySlotLookup);
```
Don't these need to be the other way around?  First put all the existing data into the map, then add the new values? Otherwise if the new blocks are from the same slot as any existing block the `putAll` will overwrite the set (whereas it should merge them).

mbaxter replied:
yes - good catch!  will make sure we have some tests around this.

ajsutton commented on storage/src/main/java/tech/pegasys/teku/storage/server/rocksdb/schema/V3Schema.java:
```diff
@@ -57,7 +57,6 @@
       RocksDbVariable.create(3, CHECKPOINT_SERIALIZER);
   RocksDbVariable<Checkpoint> FINALIZED_CHECKPOINT =
       RocksDbVariable.create(4, CHECKPOINT_SERIALIZER);
-  RocksDbVariable<BeaconState> LATEST_FINALIZED_STATE = RocksDbVariable.create(5, STATE_SERIALIZER);
```
Just confirming - this doesn't break backwards compatibility because it's not a column itself - there will just be a value stored in the DB which we no longer use?

mbaxter replied:
Right - manually tested as well to confirm.

ajsutton reviewed (commented):
Looking really good - I like how the database side is simplified and much more focussed just on storing stuff rather than implementing a bunch of the eth2 logic.

Only real concern is how slotToHotBlockRootIndex is being built in `Store`.  There are potentially other refactorings that could be done but since this is the start of a stream of work, not sure if this is the right point to try and do them or not.

mbaxter committed:
Move state generation logic into an independent utility class

mbaxter committed:
Fix bug in pruning logic, add tests

mbaxter committed:
Code review - small cleanup

mbaxter committed:
Reorganize pruning logic

mbaxter committed:
Merge branch 'master' into issue-1976/move-hot-state-mgmt-into-store

mbaxter commented on storage/src/main/java/tech/pegasys/teku/storage/Store.java:
```diff
@@ -354,23 +451,16 @@ public VoteTracker getVote(UnsignedLong validatorIndex) {
-              block_states,
-              checkpoint_states,
-              votes);
+      final TransactionCommitUpdates updates = TransactionCommitUpdates.calculate(Store.this, this);
```
Hmm - I think the current structure is problematic.  We do these calculations upfront to figure out our updates, send those updates to storage, and only then get a `writeLock` to actually update our in-memory `Store`.  But with that setup, we could interleave multiple transactions and commit the calculated changes out of order, and run into problems while trying to calculate the updates because the `Store` can change out from under us ...

mbaxter replied:
Fixed this by adding a write lock around the update calculations.  This means that our update calculations are running off of a fixed `Store`, so we don't have to worry about any data changing while we're processing.  It's possible that we could fail to prune some new blocks that end up getting committed after we run our calculations, but those data will eventually be pruned the next time we finalize a new checkpoint.

mbaxter committed:
Add CheckpointAndBlock datastructure

mbaxter committed:
Fix checkpoint ordering in StoreTest

mbaxter committed:
Throw an exception if StateGenerator fails to regenerate any state

mbaxter committed:
Add latest finalized state tracking back in

We can't substitute latest finalized state for the checkpoint state: the
checkpoint state does not necessarily correspond to the latest finalized
block's state.  If the finalized block is prior to the epoch boundary
slot, its corresponding checkpoint state will be advanced to the epoch
boundary slot.

mbaxter committed:
Simplify storage update API's

mbaxter committed:
Move Store to client package

mbaxter committed:
Lock while transaction updates are being calculated

mbaxter committed:
Move transaction update helper class into separate file

mbaxter committed:
Move logic to update Store into StoreTransactionUpdates

Fix a few issues / cleanup StoreTransactionUpdates

mbaxter committed:
Merge branch 'master' into issue-1976/move-hot-state-mgmt-into-store

mbaxter committed:
Remove dead code

ajsutton commented on ethereum/datastructures/src/main/java/tech/pegasys/teku/datastructures/state/CheckpointAndBlock.java:
```diff
@@ -0,0 +1,58 @@
+import tech.pegasys.teku.datastructures.blocks.SignedBeaconBlock;
+
+public class CheckpointAndBlock {
+  final Checkpoint checkpoint;
+  final SignedBeaconBlock block;
```
nit: private?

ajsutton commented on storage/src/main/java/tech/pegasys/teku/storage/client/Store.java:
```diff
@@ -42,27 +47,28 @@
+  Map<Bytes32, SignedBeaconBlock> blocks;
+  Map<Bytes32, BeaconState> block_states;
+  Map<Checkpoint, BeaconState> checkpoint_states;
+  Map<UnsignedLong, VoteTracker> votes;
```
It's kind of terrifying for these to not be private - there's quite a bit of stuff that could be changing them. I do like having things split into separate classes though.

Maybe we should create a dedicated `store` package so that we have a very clear scope of what can access this stuff?  Maybe could even introduce a `MutableStore` interface that most things use (which extends `ReadOnlyStore` and adds the `startTransaction` methods).  Then the actual `Store` is only ever passed to the store helper classes.

mbaxter replied:
Yeah - I was a bit uncomfortable with this as well.  Isolating the `Store` and helper classes into a separate package sounds like a good solution 👍 Will fix.

mbaxter replied:
Added some interfaces and made `Store` package-private: https://github.com/PegaSysEng/teku/pull/2006/commits/82036c69613d283ab5fef1f3ba1b4d0cad25fe6e

ajsutton commented on storage/src/main/java/tech/pegasys/teku/storage/client/Store.java:
```diff
@@ -111,6 +120,91 @@ public static Store getForkChoiceStore(final BeaconState anchorState) {
+      LOG.warn("Unable to regenerate some hot states from hot blocks");
+
+      // Drop any blocks for which a state couldn't be generated
+      new HashSet<>(Sets.difference(blocks.keySet(), blockStates.keySet())).forEach(blocks::remove);
```
Do these blocks wind up staying on disk? And is that an issue?

mbaxter replied:
I'm not throwing an error here for backwards-compatibility reasons, but we should add an error here at some point.  For existing databases, fork blocks are pruned on restart, so they could have some invalid blocks in the database that would need to be pruned here.  I'll add a comment.

ajsutton commented on storage/src/main/java/tech/pegasys/teku/storage/client/Store.java:
```diff
@@ -354,46 +460,25 @@ public VoteTracker getVote(UnsignedLong validatorIndex) {
+
       return storageUpdateChannel
-          .onStorageUpdate(updateEvent)
+          .onStorageUpdate(updates.createStorageUpdate())
```
So to think through a conversation we've had offline about whether or not the call to `storageUpdateChannel.onStorageUpdate` needs to happen inside the write lock...

 - We can only have one thread importing blocks at a time, otherwise we'd mess up the in-memory state and various block validations wouldn't be accurate
 - There might be concurrent fork choice evaluations being done which would result in other store transactions interleaving, but those would only touch votes and blocks don't touch votes.

Thus, this is safe to do outside the write lock because no other block import can start until at least after this commit method returns, by which time the storage update has been sent and has the ordering fixed from there.

By releasing the write lock, even if storage is backed up read queries are able to be performed against the store which is good.

Currently block import waits for the returned future to complete before moving on to the next block and that will make this work, but if it didn't I believe we'd have an issue here because we aren't applying the changes to the in-memory store until the write to disk has completed.

The issue is that tx 1 makes the changes for block 10, then sends them off to storage, but at that point they haven't been applied to the underlying Store so are only visible to tx 1.  If the block importer didn't wait for `commit` to complete and moved on to block 11 - block 10 wouldn't yet be visible to it and it would fail.

But if we apply the additions before we know the storage works, we're a bit hosed if the storage fails because we then can't roll back.

So we probably need to stick with doing it this way and address it if we ever want to import blocks without waiting for storage (at which point we need to handle the error case anyway). We may want to just make `commit()` synchronous since everywhere that uses it does a `.join()` anyway.

ajsutton commented on storage/src/main/java/tech/pegasys/teku/storage/client/StoreTransactionUpdates.java:
```diff
@@ -0,0 +1,251 @@
+import tech.pegasys.teku.storage.client.Store.Transaction;
+import tech.pegasys.teku.storage.events.StorageUpdate;
+
+class StoreTransactionUpdates {
```
It is really nice that this is in a separate class.

ajsutton commented on storage/src/main/java/tech/pegasys/teku/storage/client/StoreTransactionUpdates.java:
```diff
@@ -0,0 +1,251 @@
+    final CheckpointAndBlock prevFinalizedCheckpoint = baseStore.getFinalizedCheckpointAndBlock();
+    final Optional<CheckpointAndBlock> newFinalizedCheckpoint =
+        Optional.of(tx.getFinalizedCheckpointAndBlock())
+            .filter(c -> c.getEpoch().compareTo(prevFinalizedCheckpoint.getEpoch()) > 0);
```
Looks like it would be easier to not use an `Optional` here.  Just something like:
```
final CheckpointAndBlock newFinalizedCheckpoint = tx.getFinalizedCheckpointAndBlock();
...
if (newFinalizedCheckpoint.getEpoch().compareTo(prevFinalizedCheckpoint.getEpoch()) > 0) {
    ... calculate stuff to prune ...
} else {
  .. the empty map stuff ...
}
```

Although I guess you need the optional to pass into the `StoreTransactionUpdates` so happy to go either way.

mbaxter replied:
👍 Simplified

ajsutton reviewed (approved):
LGTM.

mbaxter committed:
Move Store to a dedicated package for better isolation

mbaxter committed:
Put Store behind an interface, make Store package-private

mbaxter committed:
Add explanatory comment around warning

mbaxter committed:
Merge branch 'master' into issue-1976/move-hot-state-mgmt-into-store

mbaxter committed:
Simplify finalized checkpoint handling

mbaxter committed:
Merge branch 'master' into issue-1976/move-hot-state-mgmt-into-store
